### PR TITLE
Add preview for PDF report submission

### PR DESCRIPTION
**Pull request by zaucker**
- Extend PDF report for (optional) submission
- Adapt frontend

Mail submission will be a separat PR.

### CODE REVIEW

File: frontend/source/class/agrammon/module/output/SubmitWindow.js [hunk 1]
```diff
@@ -261,42 +260,54 @@ qx.Class.define('agrammon.module.output.SubmitWindow', {
...
+                guiVariant   : agrammon.Info.getInstance().getGuiVariant(),
+                version      : this.__version,
+                mode         : 'submission',
+
+                farmSituation  : this.__farmSituationSelect.getSelection()[0].getLabel(),
+                farmNumber     : this.__farmNumberInput.getValue(),
+                comment        : comment,
+                senderName     : sender,
+                recipientName  : this.__recipientSelect.getSelection()[0].getLabel(),
+                recipientEmail : this.__recipientSelect.getSelection()[0].getModel()
```
Comment (jnthn) [inline frontend/source/class/agrammon/module/output/SubmitWindow.js:282]: Seems we're exposing an email sending endpoint here that could potentially be abused to send spam? Or do we validate the email addresses on the server side too?

File: lib/Agrammon/OutputFormatter/PDF.pm6 [hunk 2]
```diff
@@ -176,6 +177,33 @@ sub input-output-as-pdf(
...
+        submission => %(
+            farm => %(
+                :de('Betriebsnummer'),
+                :en('Farm number'),
+                :fr('fr number'),
+            ){$language},
+            situation => %(
+                :de('Situation des Betriebs'),
+                :en('Farm situation'),
+                :fr('fr situation'),
```
Comment (jnthn) [inline lib/Agrammon/OutputFormatter/PDF.pm6:189]: C'est vraiment bien francais... :-)

File: lib/Agrammon/Web/Service.pm6 [hunk 3]
```diff
@@ -163,6 +163,23 @@ class Agrammon::Web::Service {
         my $with-filters = $type eq 'reportDetailed';
         my $all-filters  = $type eq 'reportDetailed';
 
+        my %submission;
+        if %params<mode> eq 'submission' {
+            my $sender-name = %params<senderName>;
+            # frontend converts newlines to XXX
+            $sender-name ~~ s:g/XXX/\\newline\{\}/;
```
Comment (jnthn) [inline lib/Agrammon/Web/Service.pm6:170]: Hm, that feels a little odd...why?

Comment (jnthn) [reply lib/Agrammon/Web/Service.pm6:170]: And if there can be more than one newline, needs to be `s:g`

File: frontend/source/class/agrammon/module/output/SubmitWindow.js (hunk 1 above)
Comment (zaucker) [reply frontend/source/class/agrammon/module/output/SubmitWindow.js:282]: I will move the recipient address(es) to the backend config, just wanted to make the current setting work first.

File: lib/Agrammon/Web/Service.pm6 (hunk 3 above)
Comment (zaucker) [reply lib/Agrammon/Web/Service.pm6:170]: Well, it is `s:g`... I was wondering too, why I did this in the backend. I guess I can as well replace the `\n` directly in the PDF generator.

File: lib/Agrammon/OutputFormatter/PDF.pm6 (hunk 2 above)
Comment (zaucker) [reply lib/Agrammon/OutputFormatter/PDF.pm6]: Fixed :-)

File: lib/Agrammon/Web/Service.pm6 (hunk 3 above)
Comment (zaucker) [reply lib/Agrammon/Web/Service.pm6]: Fixed (and removed the replacment from frontend).

File: frontend/source/class/agrammon/module/output/SubmitWindow.js (hunk 1 above)
Comment (zaucker) [reply frontend/source/class/agrammon/module/output/SubmitWindow.js:274]: I add a check for the recipient for the moment and redirect any mail not sent to me, to me for the moment.
Will have to think about this a bit. Will probably configure the allowed addresses for submission in the config file.

There is another mail functionality for "sending" datasets to other users (just a DB copy) with an eMail notification.
There will be a check for the recipient being a valid Agrammon user. Not sure I can make that completely save against Agrammon users spamming eachother.

Comment (jnthn) [reply frontend/source/class/agrammon/module/output/SubmitWindow.js:274]: I think Agrammon users spamming each other is a much smaller (and ignorable) problem.

File: lib/Agrammon/OutputFormatter/PDF.pm6 [hunk 4]
```diff
@@ -176,6 +177,33 @@ sub input-output-as-pdf(
                 :fr('Set de données'),
             ){$language},
         ),
+        submission => %(
+            farm => %(
+                :de('Betriebsnummer'),
+                :en('Farm number'),
+                :fr("Numéro d'entreprise"),
+            ){$language},
```
Comment (jnthn) [inline lib/Agrammon/OutputFormatter/PDF.pm6:185]: Maybe some day these want to go in a `.po` file (we have a module for that), but fine for now.

File: lib/Agrammon/Web/Service.pm6 [hunk 5]
```diff
@@ -159,20 +178,39 @@ class Agrammon::Web::Service {
...
-        my $with-filters = $type eq 'reportDetailed';
-        my $all-filters  = $type eq 'reportDetailed';
+
+        my %submission;
+        if %params<mode> and %params<mode> eq 'submission' {
+            my $sender-name = %params<senderName>;
+            # replace URL encoded newlines
+            $sender-name ~~ s:g/'%D'/\\newline\{\}/;
+            my $comment = %params<comment>;
+            $comment ~~ s:g/'%D'/\\newline\{\}/;
```
Comment (jnthn) [inline lib/Agrammon/Web/Service.pm6:188]: Hm, if it's a form submission I think Cro does this automatically? Plus `%D` looks wrong, I think it's always encoded as `%` with two chars after it.

File: lib/Agrammon/OutputFormatter/PDF.pm6 (hunk 4 above)
Comment (zaucker) [reply lib/Agrammon/OutputFormatter/PDF.pm6:185]: Yes, that would make sense.

File: frontend/source/class/agrammon/module/output/SubmitWindow.js (hunk 1 above)
Comment (zaucker) [reply frontend/source/class/agrammon/module/output/SubmitWindow.js:274]: > I think Agrammon users spamming each other is a much smaller (and ignorable) problem.

Agreed.

File: lib/Agrammon/Web/Service.pm6 (hunk 5 above)
Comment (zaucker) [reply lib/Agrammon/Web/Service.pm6]: Well, I don't understand why, but this is not working as expected. See also chat. I thus reverted to doing the manual "encoding" to XXX as I am out of ideas right now.

File: dev/META6.json [hunk 6]
```diff
@@ -21,7 +21,7 @@
     "Net::SMTP::Client::Async",
     "OO::Monitors",
     "Shell::Command",
-    "Spreadsheet::XLSX:ver<0.3+>",
+    "Spreadsheet::XLSX:ver<0.2+>",
```
Comment (jnthn) [inline dev/META6.json]: The new release was `0.2.1`.

File: t/routes-dataset.t [hunk 7]
```diff
@@ -7,7 +7,7 @@ use Cro::HTTP::Test;
 use Test::Mock;
 use Test;
 
-plan 15;
+#plan 15;
```
Comment (jnthn) [inline t/routes-dataset.t:10]: Can just remove it if relying on `done-testing`.

**Review by jnthn — APPROVED** [review]

File: dev/META6.json (hunk 6 above)
Comment (zaucker) [reply dev/META6.json]: Oh, should have guessed as it was a bug fix.

File: t/routes-dataset.t (hunk 7 above)
Comment (zaucker) [reply t/routes-dataset.t:10]: In a way I prefer to set the plan.

**Pull request merged**